### PR TITLE
DocBlocks

### PR DESCRIPTION
**Pull request by bestit-lubisch**
PHPCS-17 Class must have a doc block
PHPCS-18 Empty line before every doc block
PHPCS-19 No aligning in doc comment strings
PHPCS-20 Class Doc Block summary
PHPCS-21 Class Doc Block description
PHPCS-22 Class Doc Block Author tag
PHPCS-23 Class Doc Block allowed tags
PHPCS-26 Constant / Class Property Doc Block
PHPCS-33 Method Doc Block
PHPCS-35 Method Doc Block Description
PHPCS-36 Method Doc Block Tags
PHPCS-37 Method Doc Block empty lines between tags

### CODE REVIEW

**Comment by bestit-tkellner** [comment]
Build failed. :-(

*ZONK*

File: src/Standards/BestIt/AbstractDocSniff.php [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+        $this->docHelper = new DocHelper($phpcsFile, $stackPtr);
+        $this->summaryHelper = new DocSummaryHelper($phpcsFile, $this->docHelper);
+        $this->descriptionHelper = new DocDescriptionHelper($phpcsFile, $this->docHelper, $this->summaryHelper);
+        $this->tagHelper = new DocTagHelper($phpcsFile, $this->docHelper, $stackPtr);
+
+        if (!$this->docHelper->checkCommentExists($this->listenerPtr)) {
+            return;
+        }
+
+        if (!$this->docHelper->checkCommentMultiLine()) {
```
Comment (nhardeweg) [inline src/Standards/BestIt/AbstractDocSniff.php]: Ifs zusammenfassen?

File: src/Standards/BestIt/AbstractDocSniff.php [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+     */
+    public function getTagHelper()
+    {
+        return $this->tagHelper;
+    }
+
+    /**
+     * Returns which tokens should be listened to.
+     *
+     * @return array|int List of tokens to listen for
```
Comment (nhardeweg) [inline src/Standards/BestIt/AbstractDocSniff.php]: array|int als Return?! Bitte einen Returntypen.

File: src/Standards/BestIt/Commenting/TagValidator/Validators/AuthorValidator.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+ */
+class AuthorValidator extends AbstractValidator
+{
+    /**
+     * Validates the tag content.
+     *
+     * @param int $contentPtr Pointer to the tag content
+     * @param array $contentToken Token of the tag content
+     *
+     * @return bool Indicator if content is valid or not
```
Comment (nhardeweg) [inline src/Standards/BestIt/Commenting/TagValidator/Validators/AuthorValidator.php]: I dont see any return in this method

File: src/Standards/BestIt/Commenting/TagValidator/Validators/DeprecatedValidator.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ */
+class DeprecatedValidator extends AbstractValidator
+{
+    /**
+     * Validates the tag content.
+     *
+     * @param int $contentPtr Pointer to the tag content
+     * @param array $contentToken Token of the tag content
+     *
+     * @return bool Indicator if content is valid or not
```
Comment (nhardeweg) [inline src/Standards/BestIt/Commenting/TagValidator/Validators/DeprecatedValidator.php]: Change return type to void

File: src/Standards/BestIt/AbstractDocSniff.php (hunk 2 above)
Comment (bestit-lubisch) [inline src/Standards/BestIt/AbstractDocSniff.php]: done

File: src/Standards/BestIt/Commenting/TagValidator/Validators/AuthorValidator.php (hunk 3 above)
Comment (bestit-lubisch) [inline src/Standards/BestIt/Commenting/TagValidator/Validators/AuthorValidator.php]: done.

File: src/Standards/BestIt/Commenting/TagValidator/Validators/DeprecatedValidator.php (hunk 4 above)
Comment (bestit-lubisch) [inline src/Standards/BestIt/Commenting/TagValidator/Validators/DeprecatedValidator.php]: done

File: src/Standards/BestIt/Commenting/TagValidator/Validators/PackageValidator.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+ */
+class PackageValidator extends AbstractValidator
+{
+    /**
+     * Validates the tag content.
+     *
+     * @param int $contentPtr Pointer to the tag content
+     * @param array $contentToken Token of the tag content
+     *
+     * @return bool Indicator if content is valid or not
```
Comment (nhardeweg) [inline src/Standards/BestIt/Commenting/TagValidator/Validators/PackageValidator.php]: void

File: src/Standards/BestIt/Commenting/TagValidator/Validators/ParamValidator.php [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+ */
+class ParamValidator extends AbstractValidator
+{
+    /**
+     * Validates the tag content.
+     *
+     * @param int $contentPtr Pointer to the tag content
+     * @param array $contentToken Token of the tag content
+     *
+     * @return bool Indicator if content is valid or not
```
Comment (nhardeweg) [inline src/Standards/BestIt/Commenting/TagValidator/Validators/ParamValidator.php]: void

File: src/Standards/BestIt/Commenting/TagValidator/Validators/ReturnValidator.php [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+ */
+class ReturnValidator extends AbstractValidator
+{
+    /**
+     * Validates the tag content.
+     *
+     * @param int $contentPtr Pointer to the tag content
+     * @param array $contentToken Token of the tag content
+     *
+     * @return bool Indicator if content is valid or not
```
Comment (nhardeweg) [inline src/Standards/BestIt/Commenting/TagValidator/Validators/ReturnValidator.php]: void

File: src/Standards/BestIt/Commenting/TagValidator/Validators/ThrowsValidator.php [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+ */
+class ThrowsValidator extends AbstractValidator
+{
+    /**
+     * Validates the tag content.
+     *
+     * @param int $contentPtr Pointer to the tag content
+     * @param array $contentToken Token of the tag content
+     *
+     * @return bool Indicator if content is valid or not
```
Comment (nhardeweg) [inline src/Standards/BestIt/Commenting/TagValidator/Validators/ThrowsValidator.php]: void

File: src/Standards/BestIt/Commenting/TagValidator/Validators/ValidatorInterface.php [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+ */
+interface ValidatorInterface
+{
+    /**
+     * Validates the tag content.
+     *
+     * @param int $contentPtr Pointer to the tag content
+     * @param array $contentToken Token of the tag content
+     *
+     * @return bool Indicator if content is valid or not
```
Comment (nhardeweg) [inline src/Standards/BestIt/Commenting/TagValidator/Validators/ValidatorInterface.php]: void?

File: src/Standards/BestIt/Helper/DocSummaryHelper.php [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+    {
+        $this->file = $file;
+        $this->tokens = $file->getTokens();
+        $this->docHelper = $docHelper;
+    }
+
+    /**
+     * Returns pointer to the comment summary.
+     *
+     * @return bool|int Pointer to the token or false
```
Comment (nhardeweg) [inline src/Standards/BestIt/Helper/DocSummaryHelper.php]: Dual return type.

File: src/Standards/BestIt/Helper/DocSummaryHelper.php [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+                AbstractDocSniff::MESSAGE_NO_LINE_AFTER_SUMMARY,
+                $summaryPtr,
+                AbstractDocSniff::CODE_NO_LINE_AFTER_SUMMARY
+            );
+
+            if ($fixLineAfterSummary) {
+                $this->fixNoLineAfterSummary();
+            }
+
+            return;
```
Comment (nhardeweg) [inline src/Standards/BestIt/Helper/DocSummaryHelper.php]: ??

File: src/Standards/BestIt/Helper/DocSummaryHelper.php [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+    {
+        $commentStartToken = $this->docHelper->getCommentStartToken();
+        $summaryStartToken = $this->getCommentSummaryToken();
+
+        $startLine = $commentStartToken['line'] + 1;
+        $endLine = $summaryStartToken['line'] - 1;
+
+        $this->file->fixer->beginChangeset();
+
+        foreach ($this->tokens as $tokenPtr => $token) {
```
Comment (nhardeweg) [inline src/Standards/BestIt/Helper/DocSummaryHelper.php]: FixerHelper remove line

File: src/Standards/BestIt/Helper/DocTagHelper.php [hunk 13]
```diff
@@ -0,0 +1,635 @@
+<?php
+
+namespace BestIt\Helper;
+
+use BestIt\AbstractDocSniff;
+use BestIt\Commenting\TagValidator\TagValidatorFactory;
+use PHP_CodeSniffer_File;
+use function Sodium\compare;
```
Comment (nhardeweg) [inline src/Standards/BestIt/Helper/DocTagHelper.php]: Whats this?

File: src/Standards/BestIt/Helper/DocTagHelper.php [hunk 14]
```diff
@@ -0,0 +1,635 @@
...
+
+        $this->validateTags();
+        $this->checkAllowedTags();
+        $this->checkTagOccurrences();
+
+        if (count($this->getCommentTagTokens()) === 0) {
+            return;
+        }
+
+        if (!$this->checkTagSorting()) {
```
Comment (nhardeweg) [inline src/Standards/BestIt/Helper/DocTagHelper.php]: combine ifs

File: src/Standards/BestIt/Helper/DocTagHelper.php [hunk 15]
```diff
@@ -0,0 +1,635 @@
...
+                    AbstractDocSniff::MESSAGE_TAG_OCCURENCE_MAX,
+                    $this->docHelper->getCommentStartPointer(),
+                    AbstractDocSniff::CODE_TAG_OCCURENCE_MAX,
+                    [
+                        $tagName,
+                        $max
+                    ]
+                );
+
+                continue;
```
Comment (nhardeweg) [inline src/Standards/BestIt/Helper/DocTagHelper.php]: Unnecessary

File: src/Standards/BestIt/Helper/DocTagHelper.php [hunk 16]
```diff
@@ -0,0 +1,635 @@
...
+    {
+        $tagNames = $this->getSortedTags();
+
+        return $this->getExpandedCommentTags($tagNames);
+    }
+
+    /**
+     * Returns sorted comment tag names.
+     *
+     * @return array List of current sorted comment tags
```
Comment (nhardeweg) [inline src/Standards/BestIt/Helper/DocTagHelper.php]: String[]

File: src/Standards/BestIt/Sniffs/Commenting/ConstantDocSniff.php [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+ *
+ * @package BestIt\Sniffs\Commenting
+ * @author Nick Lubisch <nick.lubisch@bestit-online.de>
+ */
+class ConstantDocSniff extends AbstractDocSniff
+{
+    /**
+     * Returns which tokens should be listened to.
+     *
+     * @return array|int List of tokens which should be listened to
```
Comment (nhardeweg) [inline src/Standards/BestIt/Sniffs/Commenting/ConstantDocSniff.php]: Dual return type.

File: src/Standards/BestIt/Sniffs/Commenting/GeneralDocSniff.php [hunk 18]
```diff
@@ -0,0 +1,343 @@
...
+     * @var string
+     */
+    const MESSAGE_NO_LINE_BEFORE_COMMENT = 'There is no line before the comment.';
+
+    /**
+     * Code that are too much lines before comment.
+     *
+     * @var string
+     */
+    const CODE_MUCH_LINES_BEFORE_COMMENT = 'MuchLinesBeforeComment';
```
Comment (nhardeweg) [inline src/Standards/BestIt/Sniffs/Commenting/GeneralDocSniff.php]: much -> multiple

File: src/Standards/BestIt/Sniffs/Commenting/GeneralDocSniff.php [hunk 19]
```diff
@@ -0,0 +1,343 @@
...
+        ];
+    }
+
+    /**
+     * Called when one of the token types that this sniff is listening for is found.
+     *
+     * @param PHP_CodeSniffer_File $phpcsFile The PHP_CodeSniffer file where the token was found.
+     * @param int $commentStartPtr The position in the PHP_CodeSniffer file's token stack where the token was found.
+     *
+     * @return void|int Optionally returns a stack pointer.
```
Comment (nhardeweg) [inline src/Standards/BestIt/Sniffs/Commenting/GeneralDocSniff.php]: int?

File: src/Standards/BestIt/Sniffs/Commenting/GeneralDocSniff.php [hunk 20]
```diff
@@ -0,0 +1,343 @@
...
+                self::MESSAGE_NO_LINE_BEFORE_COMMENT,
+                $this->commentStartPtr,
+                self::CODE_NO_LINE_BEFORE_COMMENT
+            );
+
+            if ($fixNoLine) {
+                $this->fixNoLineBeforeComment();
+            }
+
+            return;
```
Comment (nhardeweg) [inline src/Standards/BestIt/Sniffs/Commenting/GeneralDocSniff.php]: .

File: src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php [hunk 21]
```diff
@@ -1,367 +1,95 @@
...
-     * @var string
-     */
-    const MESSAGE_SPACING_AFTER = 'Additional blank lines found at end of doc comment';
-
     /**
-     * Code for missing spacing between descriptions
+     * Returns which tokens should be listened to.
      *
-     * @var string
+     * @return array|int List of tokens which is to be listened to
```
Comment (nhardeweg) [inline src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php]: Dual return type.

File: src/Standards/BestIt/Sniffs/Commenting/PropertyDocSniff.php [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * @package BestIt\Sniffs\Commenting
+ * @author Nick Lubisch <nick.lubisch@bestit-online.de>
+ */
+class PropertyDocSniff extends ConstantDocSniff
+{
+    /**
+     * Returns which tokens should be listened to.
+     *
+     * @return array|int List of tokens which is to be listened to
```
Comment (nhardeweg) [inline src/Standards/BestIt/Sniffs/Commenting/PropertyDocSniff.php]: Dual return type.

File: tests/Standards/BestIt/Sniffs/SniffTestCase.php [hunk 23]
```diff
@@ -0,0 +1,180 @@
...
+ * @author Nick Lubisch <nick.lubisch@bestit-online.de>
+ */
+abstract class SniffTestCase extends SlevomatTestCase
+{
+    /**
+     * Returns the sniff class name.
+     *
+     * @return string sniff class name
+     */
+    protected function getSniffClassName(): string
```
Comment (nhardeweg) [inline tests/Standards/BestIt/Sniffs/SniffTestCase.php]: Please either use all or no returntypes

File: tests/Standards/BestIt/Sniffs/SniffTestCase.php [hunk 24]
```diff
@@ -0,0 +1,180 @@
...
+     * Tests files with given error list and fixes them.
+     *
+     * @param string $file File to test
+     * @param string $error Error code
+     * @param int[] $lines All lines where the error code occurs
+     * @param array $sniffProperties Array of sniff properties
+     *
+     * @return void
+     */
+    protected function assertFixableErrorsInFile($file, $error, $lines, array $sniffProperties = [])
```
Comment (nhardeweg) [inline tests/Standards/BestIt/Sniffs/SniffTestCase.php]: array lines

File: src/Standards/BestIt/AbstractDocSniff.php (hunk 1 above)
Comment (bestit-lubisch) [inline src/Standards/BestIt/AbstractDocSniff.php]: done.

File: src/Standards/BestIt/Commenting/TagValidator/Validators/PackageValidator.php (hunk 5 above)
Comment (bestit-lubisch) [inline src/Standards/BestIt/Commenting/TagValidator/Validators/PackageValidator.php]: done.

File: src/Standards/BestIt/Commenting/TagValidator/Validators/ParamValidator.php (hunk 6 above)
Comment (bestit-lubisch) [inline src/Standards/BestIt/Commenting/TagValidator/Validators/ParamValidator.php]: done.

File: src/Standards/BestIt/Commenting/TagValidator/Validators/ReturnValidator.php (hunk 7 above)
Comment (bestit-lubisch) [inline src/Standards/BestIt/Commenting/TagValidator/Validators/ReturnValidator.php]: done.

File: src/Standards/BestIt/Commenting/TagValidator/Validators/ThrowsValidator.php (hunk 8 above)
Comment (bestit-lubisch) [inline src/Standards/BestIt/Commenting/TagValidator/Validators/ThrowsValidator.php]: done.

File: src/Standards/BestIt/Commenting/TagValidator/Validators/ValidatorInterface.php (hunk 9 above)
Comment (bestit-lubisch) [inline src/Standards/BestIt/Commenting/TagValidator/Validators/ValidatorInterface.php]: done.

File: src/Standards/BestIt/Commenting/TagValidator/Validators/VersionValidator.php [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+ */
+class VersionValidator extends AbstractValidator
+{
+    /**
+     * Validates the tag content.
+     *
+     * @param int $contentPtr Pointer to the tag content
+     * @param array $contentToken Token of the tag content
+     *
+     * @return bool Indicator if content is valid or not
```
Comment (bestit-lubisch) [inline src/Standards/BestIt/Commenting/TagValidator/Validators/VersionValidator.php]: done.

File: src/Standards/BestIt/Helper/DocDescriptionHelper.php [hunk 26]
```diff
@@ -0,0 +1,346 @@
...
+                    AbstractDocSniff::CODE_DESCRIPTION_TOO_LONG
+                );
+            }
+        }
+    }
+
+    /**
+     * Returns pointer to the end of the description.
+     *
+     * @return int|bool Pointer to the end of the description or false
```
Comment (bestit-lubisch) [inline src/Standards/BestIt/Helper/DocDescriptionHelper.php]: done.

File: src/Standards/BestIt/Helper/DocDescriptionHelper.php [hunk 27]
```diff
@@ -0,0 +1,346 @@
...
+            T_DOC_COMMENT_STRING,
+            $firstTagPtr - 1,
+            $descriptionStartPtr
+        );
+    }
+
+    /**
+     * Returns pointer to the start of the long description or false if not found.
+     *
+     * @return bool|int Pointer to the start of the description or false
```
Comment (bestit-lubisch) [inline src/Standards/BestIt/Helper/DocDescriptionHelper.php]: done.

File: src/Standards/BestIt/Helper/DocSummaryHelper.php (hunk 10 above)
Comment (bestit-lubisch) [inline src/Standards/BestIt/Helper/DocSummaryHelper.php]: done.

File: src/Standards/BestIt/Helper/DocSummaryHelper.php (hunk 11 above)
Comment (bestit-lubisch) [inline src/Standards/BestIt/Helper/DocSummaryHelper.php]: done.

File: src/Standards/BestIt/Helper/DocSummaryHelper.php (hunk 12 above)
Comment (bestit-lubisch) [inline src/Standards/BestIt/Helper/DocSummaryHelper.php]: refactored

File: src/Standards/BestIt/Helper/DocTagHelper.php (hunk 13 above)
Comment (bestit-lubisch) [inline src/Standards/BestIt/Helper/DocTagHelper.php]: removed

File: src/Standards/BestIt/Helper/DocTagHelper.php (hunk 14 above)
Comment (bestit-lubisch) [inline src/Standards/BestIt/Helper/DocTagHelper.php]: done.

File: src/Standards/BestIt/Helper/DocTagHelper.php (hunk 15 above)
Comment (bestit-lubisch) [inline src/Standards/BestIt/Helper/DocTagHelper.php]: done.

File: src/Standards/BestIt/Sniffs/Commenting/ConstantDocSniff.php (hunk 17 above)
Comment (bestit-lubisch) [inline src/Standards/BestIt/Sniffs/Commenting/ConstantDocSniff.php]: done.

File: src/Standards/BestIt/Sniffs/Commenting/GeneralDocSniff.php (hunk 18 above)
Comment (bestit-lubisch) [inline src/Standards/BestIt/Sniffs/Commenting/GeneralDocSniff.php]: done.

File: src/Standards/BestIt/Sniffs/Commenting/GeneralDocSniff.php (hunk 19 above)
Comment (bestit-lubisch) [inline src/Standards/BestIt/Sniffs/Commenting/GeneralDocSniff.php]: done.

File: src/Standards/BestIt/Sniffs/Commenting/GeneralDocSniff.php (hunk 20 above)
Comment (bestit-lubisch) [inline src/Standards/BestIt/Sniffs/Commenting/GeneralDocSniff.php]: done.

File: src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php (hunk 21 above)
Comment (bestit-lubisch) [inline src/Standards/BestIt/Sniffs/Commenting/MethodDocSniff.php]: done.

File: src/Standards/BestIt/Sniffs/Commenting/PropertyDocSniff.php (hunk 22 above)
Comment (bestit-lubisch) [inline src/Standards/BestIt/Sniffs/Commenting/PropertyDocSniff.php]: done.

File: tests/Standards/BestIt/Sniffs/SniffTestCase.php (hunk 24 above)
Comment (bestit-lubisch) [inline tests/Standards/BestIt/Sniffs/SniffTestCase.php]: done.

File: tests/Standards/BestIt/Sniffs/SniffTestCase.php (hunk 23 above)
Comment (bestit-lubisch) [inline tests/Standards/BestIt/Sniffs/SniffTestCase.php]: We have to talk about this next morning.
Done know the result of the last talk of this topic.

The return type is given by the interface of the Slevomat TestCase

**Comment by bestit-lubisch** [comment]
Greeeeeeen :)

**Comment by bestit-tkellner** [comment]
Commits squashen.

**Comment by bestit-lubisch** [comment]
done.

**Pull request merged**